File: name2sex/prob.pl
use Encode;

sub Train
{
	open(In,"Person.txt");
	open(Out1,">Model.txt");
	while(<In>){
		chomp;
		if(/,Î´Öª/){
			next;
		}
		
		if(/(\S+),(\S+)/){
			$Sex=$2;
			$Name=$1;
		}
		$Name=decode("gbk",$Name);
		@HZs=$Name=~/./g;
		for($i=0;$i<@HZs;$i++){
			${$Prob{$HZs[$i]}}{$Sex}++;
			$Total++;
		}
	}
	close(In);

	$Val=log(1/$Total);
	print Out1 "UNK $Val\n";
	foreach (sort keys %Prob){
			$Ref=$Prob{$_};
			$HZ=encode("gbk",$_);
			print Out1 "#$HZ\n";
			print Out2 "#$HZ\n";
			foreach (sort keys %{$Ref}){
				$Val=log(${$Ref}{$_}/$Total);
				print Out1  "$_ $Val\n";
			}	
	}

	close(Out1);
}

sub ReadModel
{
	my($File,$Ref,$RefUNK)=@_;
	open(In,"$File");
	$Line=<In>;
	if($Line=~/(\S+)\s(\S+)/){
		${$RefUNK}=$2;
	}
	while(<In>){
		chomp;
		if(/#(\S+)/){
			$HZ=$1;
		}else{
			if(/(\S+)\s(\S+)/){
				${${$Ref}{$HZ}}{$1}=$2;
			}
		}
	}
	close(In);
}

sub GetProb
{
	my($Ref,$RefUNK,$HZ,$Sex)=@_;
	if( not defined ${$Ref}{$HZ}){
		return ${$RefUNK};
	}
	
	if ( defined ${${$Ref}{$HZ}}{$Sex} ){
		return ${${$Ref}{$HZ}}{$Sex};
	}
	return ${$RefUNK};
}

sub Bayes
{
	my($Name,$Ref,$RefUNK)=@_;

	$Name=decode("gbk",$Name);
	@HZs=$Name=~/./g;
	$ProbMale=0;
	$ProbFemale=0;
	for($i=0;$i<@HZs;$i++){
		$HZ=encode("gbk",$HZs[$i]);
		$Male=GetProb($Ref,$RefUNK,$HZ,"ÄÐ");
		$Female=GetProb($Ref,$RefUNK,$HZ,"Å®");
		$ProbMale+=$Male;
		$ProbFemale+=$Female;
	}
	$Ret="Female";
	if( $ProbMale > $ProbFemale){
		$Ret="male";
	}
	return $Ret;
}

sub Main
{
	ReadModel("Model.txt",\%Model,\$UNK);

	while(1){
		print "Pls Name:";
		$Name=<stdin>;
		chomp($Name);
		$Ret=Bayes($Name,\%Model,\$UNK);	
		print "$Ret\n";
	}
}

#Train();
Main();

File: name2sex/prob.py
def ReadModel(File,RefUNK)
{
	temp=""
	In=open(File,"r")
	for Line in In:
		Line=Line.strip("#")
		Line=Line.strip()
		List=Line.split(" ")
		if not (List[1]):	#读出汉字
			if not RefUNK.get(List[0]):
				RefUNK[List[0]]={}
				temp=List[0]
				
		if (List[1]):	#读男女概率
			if not (temp):
				RefUNK["default"]=List[1]
			if not RefUNK[temp].get([List[0]]):	  #List[0]-> 男或女   List[1]-> 对应概率
				RefUNK[temp][List[0]]=List[1]				
	In.close()
}

def GetProb(hz,sex,RefUNK)
{
	if not RefUNK.get(hz):		#未找到已有汉字返回默认值
		return RefUNK["default"]
		
	if RefUNK.get(hz):
		if RefUNK[hz].get(sex):		   #汉字-性别 有数据可查时，返回数据
			return RefUNK[hz][sex]
	
	return RefUNK["default"]	  #其他情况返回默认值
}



def Bayes(name,RefUNK)
{
	male=0
	female=0
	Probmale=0
	Probfemale=0
	name1=name.decode("gbk")

	for i in range(len(name1)):
		HZ=name[i].encode("gbk")
		male=GetProb(HZ,"男",RefUNK)
		female=GetProb(HZ,"女"，RefUNK)
		Probmale+=male
		Probfemale+=female
	
	res="female"
	if(Probmale>Probfemale):
		res="male"
	
	return res
}

def Main
{
	RefUNK={}
	ReadModel("Model.txt",RefUNK);

	while 1:
	Name=""
	res=""
	Name=input("Pls Input:")
	if Name == "q":
		break
		
	Ret=Bayes(Name,RefUNK)
	print(RefUNK)
	#print(res)
	#print(Ret)
}

Main();